protocol/x/subaccounts/keeper/subaccount.go

```diff
@@ -281,12 +284,13 @@ func (k Keeper) UpdateSubaccounts(
 	}
 
 	allPerps := k.perpetualsKeeper.GetAllPerpetuals(ctx)
-	success, successPerUpdate, err = k.internalCanUpdateSubaccounts(
+	success, successPerUpdate, isolatedPerpetualPositionStateTransitions, err := k.internalCanUpdateSubaccounts(
```

Comment:
Do we really need to store the state transitions and get them from this func?
AFAIU, the 2 checks are
- opening the first position
- closing the last position

Can we just add this after the constraints check since those 2 are separate and avoid creating the whole state transition object?

Reply:
That was my initial thought as well, however it led to a lot of repeated logic for finding the perpetual updates / perpetual positions that are isolated so it was slightly cleaner to pass them through/back the functions that check constraints as both flows have very similar logic.
Also not strongly opinionated on this, as a separate function also de-couples the 2 operations (however right now they are still tightly coupled in concept).

Reply:
My preference would be to decouple the 2 operations since and err towards the simpler logic.

How about just putting this in `UpdatePerpetualPositions`? You have all the information there and the collateral transfers are more closely tied to the actual perpetual update so it might make more sense here

Reply:
I still would need some data stored that gives enough information to do the collateral transfers, and wouldn't want the collateral transfer logic to be in the same state update function.
`UpdatePerpetualPositions` could be a decent candidate for the computation, however to calculate the data for the resulting collateral transfer the `AssetUpdates` need to be unsettled, so I'd prefer doing this in the "checking" / "processing" functions before any updates to state happen and separate the state change functions from the functions calculating what state changes are needed.